### DISCUSSION TITLE
Adding additional file types

### DISCUSSION THREAD

**Discussion by DrewBold** [solspace/craft-freeform#189] [Questions & Answers] [answered]
Hello, 

How can I add additional file types to the list of allowed extensions? I have found the resource for Freeform Classic but I cannot find any documentation for Craft 3.x Freeform... 

Please advise. 

Thank you, 

Andrew McLaughlin

**Comment by kjmartens** [comment]
Hi @DrewBold,

Which file types are you wishing to add support for? Have you added the `'extraAllowedFileExtensions' => 'ai, eps, psd'` (for example) to your Craft **general.php** file? You'll then need to then allow "all" extensions for the Freeform file upload field for the form inside the form builder (uncheck all options, or check all). If I recall correctly, this would then allow all Craft defaults plus the ones you just added support for. :)

https://docs.craftcms.com/api/v3/craft-config-generalconfig.html#extraallowedfileextensions

**Comment by DrewBold** [comment]
Hello,

Yes I have added the extraAllowedFileExtensions in the general.php file 
for craft config. we can upload to the assets folder directly however 
receive an error when uploading via freeform. Please advise.

Andrew McLaughlin
Senior Developer

**Reply by kjmartens** [reply]
Can you let me know what your general.php file looks like, the extension and type of file you're attempting to upload, and the exact error that Freeform gives when uploading? :)

**Comment by DrewBold** [comment]
Hello,

Apologies for the delay, I have just returned from a little vacation. 
Please see requested information / results below. Thank you for your 
assistance.

   // Global settings
     '*' => [
         // Default Week Start Day (0 = Sunday, 1 = Monday...)
         'defaultWeekStartDay' => 1,

         'defaultTokenDuration' => 2419200,

         // Whether generated URLs should omit "index.php"
         'omitScriptNameInUrls' => true,

         // Control Panel trigger word
         'cpTrigger' => 'admin',

         // The secure key Craft will use for hashing and encrypting data
         'securityKey' => getenv('SECURITY_KEY'),

         // Whether to save the project config out to config/project.yaml
         // (see https://docs.craftcms.com/v3/project-config.html)
         'useProjectConfigFile' => false,
         'extraAllowedFileExtensions' => [
             'css','eps','csv','vtt','epub'
         ],
         //Max Upload File Size 50MB in binary bytes
         'maxUploadFileSize'=> '52428800',
     ],

Sincerely,

Andrew McLaughlin
Senior Developer

**Reply by kjmartens** [reply]
Sorry for the delay... I have a few questions:

1. Which exact versions of Craft and Freeform are you using?
2. What is the exact error you're getting in Freeform (and perhaps include a screenshot of it happening)?
3. What is the extension and application type of the file you're trying to upload?
4. Can you include a screenshot of what the file upload field's settings look like inside the form builder for this form?

Thanks! :)

**Comment by DrewBold** [comment]
Hello,

Following up on this...

Andrew McLaughlin
Senior Developer

BOLD
477 Richmond St. West, Suite 608
Toronto, Ontario
M5V 3E7
t. 416 262 3296
f. 416 504 7872

**Comment by DrewBold** [comment]
Hello,

I had provided this once previously. Here is the information as 
requested again.

- Craft CMS 3.4.28.1
- Freeform Pro 3.7.0

File extension we wish to add support for: `*.epub`

 From Global settings(Full down below):


```
'extraAllowedFileExtensions' => [
             'css','eps','csv','vtt','epub'
         ],

   // Global settings
     '*' => [
         // Default Week Start Day (0 = Sunday, 1 = Monday...)
         'defaultWeekStartDay' => 1,

         'defaultTokenDuration' => 2419200,

         // Whether generated URLs should omit "index.php"
         'omitScriptNameInUrls' => true,

         // Control Panel trigger word
         'cpTrigger' => 'admin',

         // The secure key Craft will use for hashing and encrypting data
         'securityKey' => getenv('SECURITY_KEY'),

         // Whether to save the project config out to config/project.yaml
         // (see https://docs.craftcms.com/v3/project-config.html)
         'useProjectConfigFile' => false,
         'extraAllowedFileExtensions' => [
             'css','eps','csv','vtt','epub'
         ],
         //Max Upload File Size 50MB in binary bytes
         'maxUploadFileSize'=> '52428800',
     ],
```


Andrew McLaughlin
Senior Developer

**Comment by kjmartens** [comment, marked as answer]
Thanks @DrewBold. You had previously provided what your config file looked like, but none of the info in my questions 1-4 (perhaps in a different issue or support ticket, but I need to be sure for each ticket in case it's a different install, etc). 🙂 

Anyway, I am still curious if you can answer questions 2 and 4 for me:

- What is the exact error you're getting in Freeform (and perhaps include a screenshot of it happening)?
- Can you include a screenshot of what the file upload field's settings look like inside the form builder for this form?

Further to this, are you able to upgrade to the latest version of Freeform (3.11) anytime soon? 3.7 is becoming quite old and there have been a number of fixes, changes and improvements since then. 

Finally, my hunch is that we may need to update Freeform to allow for `.epub`, as we have our own layer of file security checks happening. What you could try doing is adding a line for `epub` in the `/vendor/solspace/craft-freeform/packages/plugin/src/Library/Helpers/FileHelper.php` file in Freeform and see if that helps (assuming you've done the other steps correctly like checking off all file types for the file upload field inside the form builder):

`'application/epub+zip' => 'epub',` (assuming the mime type is correct here... it might not be - you'll need to check the file) 🙂

**Comment by DrewBold** [comment]
Apologies I have been replying through email and the attachments have not been going through to GITHUB. 
Please find the requested information below. 
![1](https://user-images.githubusercontent.com/66840687/133468531-8f163a1f-bea5-40da-a692-1125300ef49e.png)
![2](https://user-images.githubusercontent.com/66840687/133468532-69047049-fbae-4ce5-a676-c5d3eefd26fd.png)
![3](https://user-images.githubusercontent.com/66840687/133468533-5b631be0-d203-4ced-ac27-1d7d4d2cad61.png)
![4](https://user-images.githubusercontent.com/66840687/133468536-da311f5f-903e-4a7b-b64e-63155628b7ee.png)

Adding the MIME type in FileHelper.php did the trick. The upload is now working. However this is a temporary solution I feel as future updates may break this functionality. Is it possible to have the epub file type added to the supported files on a future release?

**Reply by kjmartens** [reply]
Thanks... did the exact mime type `'application/epub+zip' => 'epub',` work for you?

**Reply by DrewBold** [reply]
Yes it did. Only when allowing ALL file types to upload. This is not an ideal solution however until freeform formally supports epub files this will work.... 

When do you imagine formal support for epub files will be included in a freeform release?

**Reply by kjmartens** [reply]
I'll add that mime type to the next release. :)

**Comment by DrewBold** [comment]
@kjmartens I saw a release and did not see epub added to this release. Secondly this release removed the addition of application/epub+zip from the fileHelper.php file.....

**Reply by kjmartens** [reply]
Sorry, the fix was added to the 3.12 beta branch (so it definitely will/would have happened in time). I'll also add it to the current release now (but won't be available until 3.11.12 is released, or you could set your version to use dev-master or manually add that line back in again yourself). :)
